src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -360,9 +364,12 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
           ClassInfoType(parents, allDecls, clazz)
         } else tp
 
+
+      case tp@ClassInfoType(parents, oldDecls, clazz) if !needsMixin(clazz) => tp
```

Comment:
Thinking about this some more, we probably need to guard this with `globalPhase.isAfterRefchecks`, as we only populate the map of separated compiled superclasses in the refchecks tree transform.